### ISSUE TITLE
RangeError: Object xxx has multiple parents

### ISSUE THREAD

**Issue by johannesjo** [automerge/automerge#260]
Not sure if this is a bug or if I am missing something. The following code throws the error above.

```
doc1 = Automerge.change(doc1, 'Add card', (doc) => {
  const item = { id: 'AAA', title: 'A Rewrite everything in Clojure', done: false }
  doc.nested = {
    ids: [...doc.nested.ids, item.id],
    entities: {
      ...doc.nested.entities,
      [item.id]: item,
    },
  };
})

doc1 = Automerge.change(doc1, 'Add another card', (doc) => {
  const item = { id: 'BBB', title: 'B Rewrite everything in haha', done: false }
  doc.nested = {
    ids: [...doc.nested.ids, item.id],
    entities: {
      ...doc.nested.entities,
      [item.id]: item,
    },
  };
})
```

**Comment by ept** [comment]
@johannesjo In Automerge you cannot use this immutable object-copying pattern for updating state. If you want to add an item, you should mutate the document within the change callback, like this:

```js
doc1 = Automerge.change(doc1, 'Add card', (doc) => {
  const item = { id: 'AAA', title: 'A Rewrite everything in Clojure', done: false }
  doc.nested.ids.push(item.id)
  doc.nested.entities[item.id] = item
})
```

Automerge will take those mutations and automatically transform them into an immutable state update, so doc1 is not actually mutated.

The reason you cannot use the `{...doc.nested.entities, [item.id]: item}` idiom is that Automerge needs to track changes you make to the objects at a fine-grained level. If you replace the entire `doc.nested` object, Automerge will treat all of the contents of `doc.nested` as brand new objects, not realising that most of the contents of that object are shared with a prior state. If two users concurrently update this document, you will get a conflict at the `doc.nested` level, and none of Automerge's merging facilities work.

And the reason for the "multiple parents" error message is that you are taking an object that already exists in the document (namely one of the entities) and adding it to the same document again. This is not allowed, since each object may exist only once in the document. The state update is not able to detect that the object is simultaneously being removed from its old position.

I wonder if there is some way we could improve the error message so that we discourage people from using this `{...doc.nested.entities, [item.id]: item}` idiom.

**Comment by johannesjo** [comment]
Thank you very much for the explanation. Maybe you should add your comment as a short section to the Readme?

About the error message. Maybe it could be something a bit more verbose like:
```
Object xxx has multiple parents  – Are you directly assigning deep nested data?
```
A link to the fresh readme section would not hurt neither.

**Comment by ept** [comment]
Okay, I have added a README section and made the exception more descriptive in 4efada8928b08de50cf305e94989e629aa9880cf. Is this extra information clear?

**Comment by johannesjo** [comment]
Thank you very much! That should make things clearer for new users.
